### ISSUE TITLE
Add version tags and make "releases"

### ISSUE THREAD

**Issue by mpaluru** [freeCodeCamp/freeCodeCamp#13605]
freeCodeCamp repo currently doesn't follow any software release processes. It would be really a good idea to start versioning the software and do "releases". See: [GitHub Help: Creating Releases](https://help.github.com/articles/creating-releases)

For example: v1.0.0, v1.1.2, v2.0.1 etc. (See http://semver.org/)

Some thoughts on how this can be done:
- Pick the git hash which is currently in production (the has which is currently running on freecodecamp.com) and make it v1.0.0
- After few more commits are made, then we do another release
- We pick v1.0.1 if there are only minor bug fixes or we pick v1.1.0 if there were slightly big changes.
- If any major changes are made and they are generally not backward compatible we bump up the major number, ex: v2.0.0. An imaginary example of a major change would be moving from node.js v5 to node.js v6 or changes in mongodb version.

**Comment by erictleung** [comment]
I like the idea of versioning freeCodeCamp's codebase from a software engineering perspective. This could be useful understanding the history of changes freeCodeCamp goes through. It might even be helpful for analytics purposes.

I'm just not sure how much more overhead it will be. @BerkeleyTrue already merges changes regularly to the `master` branch. There's constant development on freeCodeCamp and versioning might be a bit hectic.

**Comment by mpaluru** [comment]
@erictleung @BerkeleyTrue I am not sure how often Berkley merges to master, but every time we merge to master that could be consider a release. From my experience working with GitHub, I don't think making a tag is a big overhead. `git tag -a v1.0 -m "Introducing version tags"` (See: https://git-scm.com/book/en/v2/Git-Basics-Tagging)

**Comment by raisedadead** [comment]
While versioning is a good idea, IMHO releases are not, as in case of other popular repositories, we are not distributing any pre-built binaries. Or as in case of JavaScript libraries, we aren't publishing to npm. In short, We are not building an app but a platform.

Sure a change log would help. Also, tagging is not the concern. Its simple process. 

But that said, the actual overhead here is the person making the releases, would have to make logs such as:

- feature X added
  - some notes
  - some more notes
- feature Y added
  - some notes
  - some more notes
- A,B & C Challenges updated to some info

You see this is a bit tedious, from our perspective, because our feature release to challenge updates ratio is quite low (almost once/twice a year to all year long).

So, our flow doesn't fit in a classical SDLC (Software Dev Life Cycle). This is going to be more of a chore. 

Also I sincerely believe that the async flow that we have lets more contributions come in. Implementing a versioning means having timelines and deadlines. Neither is not good for this project.

TL;DR:
We have enhancements all year long, and less candidates for "actual" release notes.

I would rather recommend frequent updates  on our blog/forum for the ongoing activity.

**Comment by BerkeleyTrue** [comment]
@raisedadead the dead is correct.

A website doesn't necessarily need versioning since there is only ever one version of the site running at once. While fcc is open source and there are other versions of fcc running outside of our control, this is not the a use case of the fcc site. 

I like to think of fcc as living software. 

@mpaluru thanks for opening the issue and starting the

**Comment by mpaluru** [comment]
@BerkeleyTrue @raisedadead I agree with your concerns of having additional overhead with releases. The software should at least be versioned.

Currently FCC is being deployed only on freecodecamp.com. Since it's open source, other people can and would like to setup their own version of FCC websites either for internal/external use. If there are particular version tags, the user would then be able to select one version to begin with and later update to the next stable version.

I was working on containerizing the FCC code and I was looking for ways to tag the docker image after building it. `freecodecamp:latest` is definitely the obvious tag, but this needs to point to some version at a given point.

Let's say I deploy `freecodecamp:latest` on day 1 and it was built from git hash1 ... after few more days the website was updated and `freecodecamp:latest` now points to git hash2. As a end user how would I know that `freecodecamp:latest` was updated? If there is a version bump say from v1 to v2, it makes it easier to track.

**Comment by BerkeleyTrue** [comment]
> The software should at least be versioned.

It definitely is, it's just not using something like SemVar but Git.

If it is about what to point a latest tag to it should point to master (actually backup/master until we attempt to get the new curriculum in production), which will be what is currently running.

> Let's say I deploy freecodecamp:latest on day 1 and it was built from git hash1 ... after few more days the website was updated and freecodecamp:latest now points to git hash2. As a end user how would I know that freecodecamp:latest was updated? If there is a version bump say from v1 to v2, it makes it easier to track.

This end user should assume that if the master branch has been updated then their version is out of date. 

While fCC is technically software it does not fall into the category of software that needs to communicate each individual change to the end user. (as @raisedadead pointed out above).

To the contributor, they should be working on the latest code.

To the potential third party host (which is not a use case we are supporting at the moment) they should  be using the latest code.

Our use case for containers would be to help fellow developers run fCC locally in a container so it has less of an impact on their personal environment.

**Comment by mpaluru** [comment]
Fair enough.
